### ISSUE TITLE
Exclude non-matching localities in subset LB

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#4837]
We're using the subset lb with locality weights, and ran into an issue where where requests would fail when a subset excluded an entire locality. For example:

```
L1: (50% weight)
host1: metadata: A
host2: metadata A
L2: (50% weight)
host3: metadata B
host4: metadata B
```

The subset matching A results in:
```
L1: (50% weight)
host1
host2
L2: (50% weight)
empty
```

and as a result 50% of requests fail because they get routed to L2.

A thing to note is that in the case where a subset ends up with an empty LB, this assert triggers: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L276

This is in line with the original warnings about using subset lb and locality weighting together, but in our case we know that the entire `LocalityLbEndpoint` will have the same metadata. As such I suggest one of two solutions:

1) Add a `metadata` field to `LocalityLbEndpoint` and have the subset LB exclude the *entire* locality if the metadata does not match for a given subset. This will per-locality metadata to exclude localities that do not match.

Not sure how easy this would be to implement, but wanted to see if this sounds reasonable before digging into the code too much.

2) Another option would be to special case (in the LB implementation perhaps?)  the case where a locality ends up with 0 weight and simply redistribute the locality weight of the empty locality between the remaining ones, proportional to their current weight. Without being super familiar with the EDF scheduler, this might just mean we exclude localities with 0 hosts from being inserted into the scheduler. 

A thing to note is that in the case where a subset ends up with an empty LB, this assert triggers: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L276

The ASSERT makes me think that the better solution might be to adjust the locality load (my second suggestion in the original post) so that won't be accidentally running into this ASSERT.

**Comment by snowp** [comment]
A thing to note is that in the case where a subset ends up with an empty LB, this assert triggers: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L276

This makes me think that the better solution might be to adjust the locality load in this case (maybe behind a flag?) instead of having people accidentally running into this `ASSERT`.

**Comment by snowp** [comment]
Put up https://github.com/envoyproxy/envoy/pull/4840 as an example of how 2) can be implemented

**Comment by htuch** [comment]
@snowp I'm thinking (2) sounds sane, but I wonder about edge cases. For example, imagine a locality L2 in your above example that has zero hosts in one EDS update, and then gains a single host in the next. We go from 100% traffic to L1 to 50% of traffic to L1 and 50% to L2, which seems a bit step function. Maybe this corner case isn't super interesting in your case, but just throwing it out there.

**Comment by snowp** [comment]
Yeah, we're somewhat safe from that my aligning localities with subsets. While someone could set up a locality with only one host in it, that's not specific to the subset lb. 

Would you be more comfortable if I put this behind a flag?

**Comment by snowp** [comment]
One option that comes to mind in order to smooth out the change in LB is to scale the weight of each locality with (number of hosts in filtered locality / number of hosts in original locality). Then, in the case you mentioned you get:
```
L1 weight =  50 * (2/2) = 50
L2 weight =  50 * (1/2) = 25
```

which gets you a 66/33 split between the two localities. This would work for our case since we'd always have 0 as the numerator and would help even out the load in cases where subsets do not align with localities.

This would still require the change to support handle zero weight localities to support our use case, so I'm thinking this could probably be a separate change.

**Comment by htuch** [comment]
@snowp maybe the system assigning weights should also be aware of the number of healthy endpoints in the locality and do this calculation itself? I think that would also avoid the adverse situation I described by not maintaining 50/50. Of course, that doesn't work when you have Envoys themselves HCing vs. a centralized HC and health status via EDS. Which situation do you have?

**Comment by snowp** [comment]
We have Envoys health checking. 

Don't we already account for % of health hosts in a locality here https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L271 ?

Sounds like we're talking about two different issues: scaling locality weight based on host *health*, and scaling locality weight based on host *presence* due to being filtered out by the subset LB.

I think the health case is dealt with in Envoy already by the code I linked (unless I'm misunderstanding). It might be nice to match this logic for subsets (maybe make it opt-in?), though I'm not sure if including an overprovisioning factor makes sense for subsets.

**Comment by htuch** [comment]
@snowp yeah, that actually seems pretty nice, I misunderstood in the above reply. Do any other @envoyproxy/maintainers or other community members want to opine before we commit?

**Comment by snowp** [comment]
Happy to do the work for this if we're happy with the design, should be fairly straightforward. Would we want to guard this behind a configuration flag? It sounds like a sane default, but there might people relying on the existing behavior?

**Comment by htuch** [comment]
AFAIK, you're the only folks who have been specifically interesting in the intersection of subset LB and priorities, @zuercher do you know who else might be using this?

**Comment by zuercher** [comment]
cc @rgs1, who is using subset lb iirc

**Comment by rgs1** [comment]
@zuercher thanks for tagging, looking.

**Comment by rgs1** [comment]
@htuch we (Pinterest) are interested too

**Comment by htuch** [comment]
@rgs1 can you comment on how the above proposal would affect you?

**Comment by rgs1** [comment]
cc: @fishcakez as well

**Comment by rgs1** [comment]
@htuch @snowp so scaling locality weight based on host presence due to subset LB matching sounds reasonable. Making it configurable also sounds reasonable (we could start with off by default, and then switch if over once we've gained confidence).

**Comment by rgs1** [comment]
@snowp how would the `default_subset` policy affect this?

**Comment by snowp** [comment]
The `default_subset` itself is just another subset, so I'd imagine it would work exactly the same way as the subsets targeted using route metadata (ie it would also have its weights scaled by host presence)

**Comment by rgs1** [comment]
@snowp sorry, i meant the `fallback_policy` when using `ANY_ENDPOINT` — would that mean disabling scaling locality based on subset LB matching?

**Comment by snowp** [comment]
@rgs1 `ANY_ENDPOINT` is just a subset with a predicate that matches anything, so the resulting weights should be unaffected by the proposal. Each locality will retain all its hosts, so no adjustment is made.

**Comment by snowp** [comment]
Sounds like the consensus here is that this seems like we should make this configurable (for now at least). I'll try to find the time to get this done sometime this week.
